latitudesh/lsh
# feat: Add Tags to CLI

### What does this PR do?
This PR aims to implement tag functionality to the CLI

### How
- [x]  manage tags in other resources (servers, projects, virtual networks and ssh keys).
- [x] create a `tags` command to manage the tag resource.
- [x] refactor parts of the code to prepare for code generation (https://github.com/latitudesh/lsh/pull/41)

The tags command is implemented following a structure that will make it easier both to maintain and to generate code. The objective is to migrate all other commands to this structure using a code generator in the near future.

All Tags resource code should reside inside the `cmd/tags` folder, with a `build_tags.go` file used to create the top-level tags command and attach it to the root.

The `cmd/lsh` folder contains code that should be globally accessible by the CLI, allowing us to avoid circular imports.

### How to test
Now the main.go file resides at the root folder so to build the CLI you should run:
```
go build -o lsh
```
Then you can test the tag resource:
```
./lsh tags list
./lsh tags create
./lsh tags update
./lsh tags destroy
``` 
You can also verify the tag functionality in other resources
```
# Add tags
./lsh <resource> update --tags <tag-id-1>,<tag-id-2>

# Filter by tags
./lsh <resource> list --tags <tag-id-1>,<tag-id-2>

# Remove tags:
./lsh <resource> update --tags "" 
``` 
where `<resource>` can be `servers`, `virtual_networks`, `projects` or `ssh_keys`
